### PR TITLE
Km web3module

### PR DESCRIPTION
**Pull request by thirtyaughtsix**

### CODE REVIEW

File: src/components/listing/FileLister.vue [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      await listing.list(listModule.listing.hash)
+      listModule.setPercentComplete(100)
+      listModule.setStatus(ProcessStatus.Complete)
+      ffaListingsModule.addUpload(listModule.listing)
+    } catch {
+      listModule.setStatus(ProcessStatus.Error)
+    }
+  }
+}
+</script>
```
Comment (robrobbins) [inline src/components/listing/FileLister.vue]: what does this have to do with web3?

Comment (robrobbins) [reply src/components/listing/FileLister.vue]: ic. looks fine

File: src/components/ui/Navigation.vue [hunk 2]
```diff
@@ -16,7 +16,7 @@
           <router-link to="/">Support</router-link>
         </div>
         <div class="navbar-item connect">
-          <a href="" v-on:click="setAddress">Connect</a>
+          <a href="" v-on:click="setPublicWalletAddress">Connect</a>
```
Comment (robrobbins) [inline src/components/ui/Navigation.vue]: PublicWalletAddress is some new word. i hate new words. let's decide what to call the public key and use it everywhere. 
* Wallet is overloaded and meaningless
* Address does suck
* PublicKey is nice...

File: src/components/ui/Navigation.vue [hunk 3]
```diff
@@ -35,26 +35,40 @@ import { getModule } from 'vuex-module-decorators'
...
-    } else {
-      metaMaskModule.setAddress(result)
-      const flash = new Flash(Messages.METAMASK_CONNECTED, FlashType.info)
-      flashesModule.append(flash)
+    const accept = typeof result === 'string'
+
+    let message = Errors.METAMASK_NOT_CONNECTED
+    let flashType = FlashType.error
+
+    switch (accept) {
```
Comment (robrobbins) [inline src/components/ui/Navigation.vue]: switch should not be used for a binary (true/false). Use an `if` here

or use a ternary in your let's above: `const message = accept ? Messages.Foo : Errors.Bar`

File: src/models/protocol/Voting.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    const vote = isYea ? 1 : 0
+    const method =  await this.voting.vote(listingHash, vote, this.transactionOptions)
+    const unsigned = await buildTransaction(this.web3, method)
+    await ethereum.send(unsigned)
+  }
+
+  public async getGas(methodName: string): Promise<number> {
+    return await this.voting.getGas(methodName)
+  }
+}
```
Comment (robrobbins) [inline src/models/protocol/Voting.ts]: make sure to review this pattern you are establishing with @MelvinMallari

File: src/modules/MetaMaskModule.ts [hunk 5]
```diff
@@ -3,10 +3,10 @@ import MetaMask from '../models/MetaMask'
 
 @Module({ namespaced: true, name: 'metaMaskModule' })
 export default class MetaMaskModule extends VuexModule {
-  public address: string = ''
+  public publicWalletAddress: string = ''
 
   @Mutation
-  public setAddress(address: string) {
-    this.address = address
+  public setPublicWalletAddress(publicWalletAddress: string) {
```
Comment (robrobbins) [inline src/modules/MetaMaskModule.ts]: setPublicKey maybe? I'll let you and @MelvinMallari make the actual decision

File: src/modules/UploadModule.ts [hunk 6]
```diff
@@ -1,16 +1,19 @@
...
-  VuexModule} from 'vuex-module-decorators'
+  VuexModule,
+  getModule } from 'vuex-module-decorators'
 import FfaProcessModule from '../interfaces/vuex/FfaProcessModule'
 import { ProcessStatus } from '../models/ProcessStatus'
-import FileTypeHelper from '../util/FileHelper'
 import FileHelper from '../util/FileHelper'
+import MetaMaskModule from './MetaMaskModule'
+import { Errors } from '../util/Constants'
+import { keccak256 } from 'js-sha3'
```
Comment (robrobbins) [inline src/modules/UploadModule.ts]: no, you shouldn't need another lib. Web3 has keccak

File: src/modules/UploadModule.ts [hunk 7]
```diff
@@ -119,6 +115,20 @@ export default class UploadModule extends VuexModule implements FfaProcessModule
     return FileHelper.fileSizeString(this.file.size)
   }
 
+  get hash(): string {
+    const metaMaskModule = this.context.rootState.metaMaskModule as MetaMaskModule
+    if (metaMaskModule.publicWalletAddress.length === 0) {
+      throw new Error(Errors.ADDRESS_EMPTY)
```
Comment (robrobbins) [inline src/modules/UploadModule.ts]: let's just ditch `address` everywhere

File: src/modules/UploadModule.ts [hunk 8]
```diff
@@ -119,6 +115,20 @@ export default class UploadModule extends VuexModule implements FfaProcessModule
...
+    }
+    if (this.title.length === 0) {
+      throw new Error(Errors.TITLE_EMPTY)
+    }
+    const hashedAccount = keccak256(metaMaskModule.publicWalletAddress)
+    const hashedTitle = keccak256(this.title)
+    const hash = keccak256(`${hashedAccount}${hashedTitle}`)
+    return `0x${hash}`
```
Comment (robrobbins) [inline src/modules/UploadModule.ts]: make sure 0x isn't already there. i remember i changed hashing libs once and ended up with `0x0x...`

File: src/modules/Web3Module.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  Module,
+  VuexModule,
+  Mutation } from 'vuex-module-decorators'
+import Web3 from 'web3'
+import { HttpProvider } from 'web3/providers'
+
+@Module({ namespaced: true, name: 'web3Module' })
+export default class Web3Module extends VuexModule {
+
+  public web3: Web3 = Object()
```
Comment (robrobbins) [inline src/modules/Web3Module.ts]: is the throw-away val needed?

File: src/util/Constants.ts [hunk 10]
```diff
@@ -16,6 +16,8 @@ export const Errors = {
   UPLOAD_FAILED: 'Upload failed',
   LISTING_FAILED: 'Listing failed',
   VOTING_FAILED: 'Voting failed',
+  ADDRESS_EMPTY: 'Address cannot be empty',
```
Comment (robrobbins) [inline src/util/Constants.ts]: address === bad

File: tests/unit/modules/UploadModule.spec.ts [hunk 11]
```diff
@@ -107,10 +104,20 @@ describe('UploadModule.ts', () => {
...
     expect(uploadModule.md5).toEqual('')
     expect(uploadModule.status).toBe(ProcessStatus.NotReady)
   })
+
+  it ('correctly generates hashes', () => {
+    const metaMaskModule = getModule(MetaMaskModule, appStore)
+    metaMaskModule.setPublicWalletAddress('address')
+    const uploadModule = getModule(UploadModule, appStore)
+    uploadModule.setTitle('title')
+    expect(uploadModule.hash).toEqual('0x52a6ac5d523467bd8af28d0067f695048277b504b34ae96525f5c3300644d5cd')
```
Comment (robrobbins) [inline tests/unit/modules/UploadModule.spec.ts]: ugh. hard coded hash... likely better to just check the `.length` as any keccak is guaranteed in len...

Comment (robrobbins) [reply tests/unit/modules/UploadModule.spec.ts]: you can also test verses the Datatrust contract in .js
`datatrust.getHash(str)` will return a keccak

File: src/modules/Web3Module.ts (hunk 9 above)
Comment (thirtyaughtsix) [reply src/modules/Web3Module.ts]: The Typescript vuex modules need the property initialized with a non-null/non-undefined value to work correctly.  Stuff like `public web3!: Web3` doesn't work - run into state management problems later.  Still looking into it...

File: src/components/ui/Navigation.vue (hunk 3 above)
Comment (thirtyaughtsix) [reply src/components/ui/Navigation.vue]: So this is something I've been playing with trying to use a more pattern-matching approach.  It does seem heavy handed for simple ifs.

File: src/modules/UploadModule.ts (hunk 6 above)
Comment (thirtyaughtsix) [reply src/modules/UploadModule.ts]: ugh.  it's such a pain in the ass to get to their keccak.

File: src/components/ui/Navigation.vue (hunk 2 above)
Comment (MelvinMallari) [reply src/components/ui/Navigation.vue]: Think we should stick syntactically consistent with using the shorthand for all `v-on:`'s and just use `@`'s instead.

File: src/components/listing/FileLister.vue (hunk 1 above)
Comment (thirtyaughtsix) [reply src/components/listing/FileLister.vue]: how do i put emojis in here?

File: src/components/ui/Navigation.vue (hunk 2 above)
Comment (thirtyaughtsix) [reply src/components/ui/Navigation.vue]: changing them all to publicKey

File: src/modules/UploadModule.ts (hunk 7 above)
Comment (thirtyaughtsix) [reply src/modules/UploadModule.ts]: done

File: tests/unit/modules/UploadModule.spec.ts (hunk 11 above)
Comment (thirtyaughtsix) [reply tests/unit/modules/UploadModule.spec.ts]: calling the Datatrust contract in the spec is a call too far for me.  expecting len > 2 and testing `startsWith`

File: src/models/protocol/Voting.ts (hunk 4 above)
Comment (thirtyaughtsix) [reply src/models/protocol/Voting.ts]: i've been implementation with some suspicion.  it's been copy-pasta-ed to the other HOCHOCs.  I suspect they may all get nuked.  Will take a closer inspection in another PR.

File: src/components/ui/Navigation.vue [hunk 12]
```diff
@@ -16,7 +16,7 @@
           <router-link to="/">Support</router-link>
         </div>
         <div class="navbar-item connect">
-          <a href="" v-on:click="setAddress">Connect</a>
+          <a href="" v-on:click="setPublicKey">Connect</a>
```
Comment (MelvinMallari) [inline src/components/ui/Navigation.vue]: Would be nice to be syntactically consistent in using the shorthand `@` instead of `v-on:` throughout the app

Comment (thirtyaughtsix) [reply src/components/ui/Navigation.vue]: that's a good catch!

File: src/components/listing/FileLister.vue [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+@Component
+export default class FileLister extends Vue {
+
+  public mounted(this: FileLister) {
+    this.$store.subscribe(this.vuexSubscriptions)
+  }
+
+  private vuexSubscriptions(mutation: MutationPayload, state: any) {
+    switch (mutation.type) {
```
Comment (robrobbins) [inline src/components/listing/FileLister.vue]: if there are more options coming thru here later and you are being future-friendly.... OK

if not kill this *switchception* and use good ol `if/else`

Comment (thirtyaughtsix) [reply src/components/listing/FileLister.vue]: should be moar coming.
